### DISCUSSION TITLE
Help for DWG/DXF to Pdf app

### DISCUSSION THREAD

**Discussion by DomCR** [DomCR/ACadSharp#96] [Q&A] [answered]
https://github.com/DomCR/ACadSharp/discussions/89#discussioncomment-4804025

@LeonidKaralis1993 please reply into this discussion about your topic.

**Comment by DomCR** [comment]
@LeonidKaralis1993

> Hello!
> Thanks for your teaching.
> Now, I have problem to get the string of MTEXT.
> I can't get the correct string values from MTEXT entities, I have got "SEE" to use 'Value' property .
> What's the reason?

For this issue you can open an issue into this repository to treat this specific matter, please give detailed information and some examples of the MText that you cannot read, if you can give an specific dwg file to test it that would be very helpful.

**Reply by chakra377** [reply]
Thanks for your help!
I've created a repository https://github.com/LeonidKaralis1993/ACAD_C-.git .
It contains "sample_files" folder in root directory.

**Reply by DomCR** [reply]
please review the repo that you send, there is no **sample_files** folder

**Reply by chakra377** [reply]
Have added sample_files folder.

**Comment by chakra377** [comment, marked as answer]
Could you please elaborate on the problem below?
I don't understand AtuoCad Entities, so I'm having a hard time passing him to PDF.
Please send me an explanation of his geometrical properties.
---ARC, LWPOLYLINE, HATCH, INSERT, DIMENSION_LINEAR---
Thank you.

**Reply by DomCR** [reply]
All the entities have documentation in the code about the different properties, I can summarize which ones have the geometric values:

**ARC**:
- Center
- StartAngle
- EndAngle

**LWPOLYLINE**:
- Vertices (is a list of vertices in which each one has a Location)

**DIMENSION_LINEAR**:
- FirstPoint 
- SecondPoint 

**INSERT**:
- InsertPoint 
Insert is a reference to a block, so you will need to access to the property Block and check the different entities in it

**HATCH**:
- Paths (is a list with the different boundary paths that form this hatch, each one has it's own geometric definitions)

**Reply by chakra377** [reply]
Thanks very much!
In ARC entity,  I insert a code as following.
                           if (e.ObjectType.ToString() == "ARC")
                            {
                                Console.Write("\n\t\t--DrawArc--");
                                arc = (Arc)e;
                                xRect.X = arc.Center.X + arc.Radius;
                                xRect.Y = vh - arc.Center.Y - arc.Radius;
                                xRect.Width = arc.Radius * 2;
                                xRect.Height = arc.Radius * 2;
                                Console.WriteLine(arc.Radius + " " + arc.Center);
                                gfx.DrawArc(pdfpen, xRect, arc.StartAngle, arc.EndAngle);
                            }
but the result is a line. 
Please teach me about this.
this is my result pdf.
[mySample.zip](https://github.com/DomCR/ACadSharp/files/10544372/mySample.zip)

**Reply by DomCR** [reply]
Sorry but I don't know how `PDFsharp` works or handles drawing arcs I cannot help with this matter.

**Reply by chakra377** [reply]
Resolve this problem thanks for your help.
the code is following
                            if (e.ObjectType.ToString() == "ARC")
                            {
                                Console.Write("\n\t\t--DrawArc--");
                                double stA,edA;
                                double goA;
                                arc = (Arc)e;
                                xRect.X = arc.Center.X - arc.Radius + 200;
                                xRect.Y = vh - arc.Center.Y - arc.Radius - 200;
                                xRect.Width = arc.Radius * 2;
                                xRect.Height = arc.Radius * 2;

                                if (arc.EndAngle < arc.StartAngle)
                                {
                                    stA = 360 - arc.StartAngle / pi * 180;
                                    gfx.DrawArc(pdfpen, xRect, 0, stA);
                                    edA = 360 - arc.EndAngle / pi * 180;
                                    gfx.DrawArc(pdfpen, xRect, edA, 359.999999 - edA);
                                    
                                }
                                else {
                                    stA = 360 - arc.EndAngle / pi * 180;
                                    goA = 360 - arc.StartAngle / pi * 180 - stA;
                                    gfx.DrawArc(pdfpen, xRect, stA, goA);          
                               }     
                                
                            }

**Comment by chakra377** [comment]
Thank you. DomCR!!!
Please let me know about the DIMENSION_LINEAR---FirstPoint & SecondPoint in more detail.
And How to get the "string" in METXT entity.
I used AdditionalText and got a string called "SEE".
What is the cause?

**Reply by DomCR** [reply]
DIMENSION_LINEAR is as it says, is a line, so you have the first and second point.

About the MTEXT I would need an example to be able to reproduce the error that you are having.

**Reply by DomCR** [reply]
For more information about the entities check https://help.autodesk.com/view/OARX/2021/ENU/?guid=GUID-7D07C886-FD1D-4A0C-A7AB-B4D21F18E484 this is the documentation that `ACadSharp` works with.

**Reply by chakra377** [reply]
Thank you! That's a good point for my project.

**Comment by chakra377** [comment]
Thank you.
How to get the coordinate properties of DWG/DXF files.
I am trying to get the information of the coordinate system and rearrange it on the PDF page, but the property of the coordinate system of the AutoCAD file is not known.

**Reply by DomCR** [reply]
Check the `CadHeader` in the document, you should find the System Variables from Autocad.

**Reply by chakra377** [reply]
Please tell me the name of the variable.
I'd be very grateful if you could give me a way to get the name of a variable.
Thank you!

**Comment by chakra377** [comment]
Thanks for your help
Where is the starting point for ARC's startAngle in AutoCad?
-> Where is the point in the area where the ARC pointed to by   startAngle=0    is displayed?

**Reply by DJGosnell** [reply]
Answering all questions related to DWG implementation details would require a substantial amount of time due to their intricate nature. The official specifications are documented in the DXF reference (found at https://github.com/DomCR/ACadSharp/blob/master/reference/autocad_2012_pdf_dxf-reference_enu.pdf) and the less formal DWG references (found at https://github.com/DomCR/ACadSharp/blob/master/reference/OpenDesign_Specification_for_.dwg_files.pdf). For a comprehensive understanding, I would advise reviewing these resources, as the specifications can become quite complex. The information on arcs can be found on page 66 of the DXF documentation.

It is important to note that various libraries may have different methods of representing arcs and ellipses, and conversion may be necessary to match the representation used by your library. Ellipses, in particular, can pose challenges.

**Reply by chakra377** [reply]
thank you.

**Reply by chakra377** [reply]
Have Sucess! Thank you. 
Your help has played a big role.
                            if (e.ObjectType.ToString() == "ARC")
                            {
                                Console.Write("\n\t\t--DrawArc--");
                                double stA,edA;
                                double goA;
                                arc = (Arc)e;
                                xRect.X = arc.Center.X - arc.Radius + 200;
                                xRect.Y = vh - arc.Center.Y - arc.Radius - 200;
                                xRect.Width = arc.Radius * 2;
                                xRect.Height = arc.Radius * 2;

                                if (arc.EndAngle < arc.StartAngle)
                                {
                                    stA = 360 - arc.StartAngle / pi * 180;
                                    gfx.DrawArc(pdfpen, xRect, 0, stA);
                                    edA = 360 - arc.EndAngle / pi * 180;
                                    gfx.DrawArc(pdfpen, xRect, edA, 359.999999 - edA);
                                    
                                }
                                else {
                                    stA = 360 - arc.EndAngle / pi * 180;
                                    goA = 360 - arc.StartAngle / pi * 180 - stA;
                                    gfx.DrawArc(pdfpen, xRect, stA, goA);          
                               }     
                                
                            }

**Comment by chakra377** [comment]
Thank you, everyone.
*LWPOLYLINE, HATCH, INSERT, DIMENSION_LINEAR*
A comprehensive explanation of how to draw with properties is needed.
Just sending me the URL would be greatly appreciated.

**Comment by chakra377** [comment]
Thank you.
I am trying to convert DXF files, but an error occurred in the reading function.
Please explain the differences between DXF and DWG files and the main points in drawing.

**Comment by chakra377** [comment]
The project under development is almost complete.
Please look at the generated PDF file and give me your advice.
